File: integrations/java/oicana/src/test/java/com/oicana/FontsTest.java
package com.oicana;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for host fonts registered through {@link Configuration}.
 */
class FontsTest {

    private static String manifestRequiring(String family) {
        return """
                [package]
                name = "font-test"
                version = "0.1.0"
                entrypoint = "main.typ"

                [tool.oicana]
                manifest_version = 1

                [tool.oicana.fonts]
                require = ["%s"]
                """.formatted(family);
    }

    private static final String PLAIN_MANIFEST = """
            [package]
            name = "font-test"
            version = "0.1.0"
            entrypoint = "main.typ"

            [tool.oicana]
            manifest_version = 1
            """;

    private static byte[] packTemplate(String manifest, String mainTypst) throws IOException {
        var stream = new ByteArrayOutputStream();
        try (var zip = new ZipOutputStream(stream)) {
            zip.setMethod(ZipOutputStream.STORED);
            for (var entry : Map.of("typst.toml", manifest, "main.typ", mainTypst).entrySet()) {
                byte[] content = entry.getValue().getBytes(StandardCharsets.UTF_8);
                var zipEntry = new ZipEntry(entry.getKey());
                zipEntry.setSize(content.length);
                var crc = new java.util.zip.CRC32();
                crc.update(content);
                zipEntry.setCrc(crc.getValue());
                zip.putNextEntry(zipEntry);
                zip.write(content);
                zip.closeEntry();
            }
        }
        return stream.toByteArray();
    }

    /**
     * Family the test font provides. No system or Typst-embedded font has it, so a
     * template requiring it can only be registered once the host registers the font.
     */
    private static final String TEST_FAMILY = "Oicana Test";

    /** The test font shipped with the repository. */
    private static Path aFontFile() {
        return Path.of("../../../assets/fonts/oicana-test-font.ttf").toAbsolutePath().normalize();
    }

    // The font registry is process-global, so isolate every test.
    @BeforeEach
    void clearBefore() {
        Configuration.clearFonts();
    }

    @AfterEach
    void clearAfter() {
        Configuration.clearFonts();
    }

    @Test
    void registryStartsEmpty() {
        assertEquals(List.of(), Configuration.registeredFonts());
    }

    @Test
    void registersFontsFromBytesWithoutAPath() throws IOException {
        byte[] data = Files.readAllBytes(aFontFile());

        assertEquals(1, Configuration.registerFonts(data));

        var fonts = Configuration.registeredFonts();
        assertEquals(1, fonts.size());
        assertEquals(TEST_FAMILY, fonts.get(0).family());
        // Registered from memory, so no path is reported.
        assertNull(fonts.get(0).path());
        assertTrue(fonts.get(0).file().isEmpty());
    }

    @Test
    void dataWithoutAFontIsIgnored() {
        assertEquals(0, Configuration.registerFonts("not a font".getBytes(StandardCharsets.UTF_8)));
        assertEquals(List.of(), Configuration.registeredFonts());
    }

    @Test
    void registersFontsByPathAndReportsThePath() {
        Path path = aFontFile();

        assertEquals(1, Configuration.registerFontPaths(path.toString()));

        var fonts = Configuration.registeredFonts();
        assertEquals(1, fonts.size());
        assertEquals(TEST_FAMILY, fonts.get(0).family());
        assertEquals(path.toString(), fonts.get(0).path());
        assertTrue(fonts.get(0).file().isPresent());
    }

    @Test
    void unreadablePathsAreSkipped() {
        assertEquals(0, Configuration.registerFontPaths("/nonexistent/font.ttf"));
        assertEquals(List.of(), Configuration.registeredFonts());
    }

    @Test
    void clearFontsEmptiesTheRegistry() {
        Configuration.registerFontPaths(aFontFile().toString());
        assertFalse(Configuration.registeredFonts().isEmpty());

        Configuration.clearFonts();

        assertEquals(List.of(), Configuration.registeredFonts());
    }

    @Test
    void templateRequiringAnUnavailableFamilyIsRejected() throws IOException {
        byte[] templateFile = packTemplate(manifestRequiring("Nonexistent Host Family"), "Content");

        var error = assertThrows(OicanaException.class, () -> new Template(templateFile));

        assertTrue(error.getMessage().contains("Nonexistent Host Family"));
    }

    @Test
    void testTemplateIsRejectedUntilTheFontIsRegistered() throws IOException {
        byte[] templateFile = packTemplate(manifestRequiring(TEST_FAMILY), "Content");

        // Proves the family really is unavailable without the host font.
        var error = assertThrows(OicanaException.class, () -> new Template(templateFile));

        assertTrue(error.getMessage().contains(TEST_FAMILY));
    }

    @Test
    void templateRequiringARegisteredFamilyCompiles() throws IOException {
        Configuration.registerFontPaths(aFontFile().toString());

        byte[] templateFile = packTemplate(manifestRequiring(TEST_FAMILY), "Content");

        try (var template = new Template(templateFile)) {
            byte[] svg = template.exportSvg(Map.of(), Map.of(), CompilationMode.DEVELOPMENT, null);

            assertTrue(new String(svg, StandardCharsets.UTF_8).contains("<svg"));
        }
    }

    @Test
    void registeredFontRendersWithoutAWarning() throws IOException {
        Configuration.registerFontPaths(aFontFile().toString());

        byte[] templateFile = packTemplate(
                PLAIN_MANIFEST,
                "#set text(font: \"%s\")\nContent".formatted(TEST_FAMILY));

        try (var template = new Template(templateFile)) {
            template.exportSvg(Map.of(), Map.of(), CompilationMode.DEVELOPMENT, null);

            assertTrue(template.warnings().isEmpty());
        }
    }

    /** Without the host font, the same template falls back and warns. */
    @Test
    void unregisteredFamilyWarns() throws IOException {
        byte[] templateFile = packTemplate(
                PLAIN_MANIFEST,
                "#set text(font: \"%s\")\nContent".formatted(TEST_FAMILY));

        try (var template = new Template(templateFile)) {
            template.exportSvg(Map.of(), Map.of(), CompilationMode.DEVELOPMENT, null);

            assertTrue(template.warnings().isPresent());
            assertTrue(template.warnings().get().contains(TEST_FAMILY));
        }
    }
}

File: integrations/php/oicana-php/tests/FontsTest.php
<?php

declare(strict_types=1);

use Oicana\CompilationMode;
use Oicana\Configuration;
use Oicana\ExportFormat;
use Oicana\RegisteredFont;
use Oicana\Template;

beforeEach(function () {
    if (!extension_loaded('oicana')) {
        test()->markTestSkipped('oicana extension not loaded');
    }

    // The font registry is process-global, so isolate every test.
    Configuration::clearFonts();
});

afterEach(function () {
    Configuration::clearFonts();
});

/**
 * Family the test font provides. No system or Typst-embedded font has it, so a
 * template requiring it can only be registered once the host registers the font.
 */
const TEST_FAMILY = 'Oicana Test';

/** The test font shipped with the repository. */
function a_font_file(): string
{
    return dirname(__DIR__, 4) . '/assets/fonts/oicana-test-font.ttf';
}

/** Pack a template whose manifest requires the given font family. */
function pack_template_requiring(string $family): string
{
    $manifest = <<<TOML
        [package]
        name = "font-test"
        version = "0.1.0"
        entrypoint = "main.typ"

        [tool.oicana]
        manifest_version = 1

        [tool.oicana.fonts]
        require = ["{$family}"]
        TOML;

    $path = tempnam(sys_get_temp_dir(), 'oicana-font-test-') . '.zip';
    $zip = new ZipArchive();
    $zip->open($path, ZipArchive::CREATE | ZipArchive::OVERWRITE);
    $zip->addFromString('typst.toml', $manifest);
    $zip->addFromString('main.typ', 'Content');
    $zip->setCompressionName('typst.toml', ZipArchive::CM_STORE);
    $zip->setCompressionName('main.typ', ZipArchive::CM_STORE);
    $zip->close();

    $bytes = file_get_contents($path);
    assert(is_string($bytes));
    unlink($path);

    return $bytes;
}

test('registry starts empty', function () {
    expect(Configuration::registeredFonts())->toBe([]);
});

test('fonts can be registered from bytes without a path', function () {
    $data = file_get_contents(a_font_file());
    assert(is_string($data));

    expect(Configuration::registerFonts([$data]))->toBe(1);

    $fonts = Configuration::registeredFonts();
    expect($fonts)->toHaveCount(1);
    expect($fonts[0])->toBeInstanceOf(RegisteredFont::class);
    expect($fonts[0]->family)->toBe(TEST_FAMILY);
    // Registered from memory, so no path is reported.
    expect($fonts[0]->path)->toBeNull();
});

test('data without a font is ignored', function () {
    expect(Configuration::registerFonts(['not a font']))->toBe(0);
    expect(Configuration::registeredFonts())->toBe([]);
});

test('fonts registered by path report the path', function () {
    $path = a_font_file();

    expect(Configuration::registerFontPaths([$path]))->toBe(1);

    $fonts = Configuration::registeredFonts();
    expect($fonts)->toHaveCount(1);
    expect($fonts[0]->family)->toBe(TEST_FAMILY);
    expect($fonts[0]->path)->toBe($path);
});

test('unreadable paths are skipped', function () {
    expect(Configuration::registerFontPaths(['/nonexistent/font.ttf']))->toBe(0);
    expect(Configuration::registeredFonts())->toBe([]);
});

test('clearFonts empties the registry', function () {
    Configuration::registerFontPaths([a_font_file()]);
    expect(Configuration::registeredFonts())->not->toBe([]);

    Configuration::clearFonts();

    expect(Configuration::registeredFonts())->toBe([]);
});

test('a template requiring an unavailable family is rejected', function () {
    $template = pack_template_requiring('Nonexistent Host Family');

    expect(fn () => new Template($template))
        ->toThrow(Exception::class, 'Nonexistent Host Family');
});

test('the test template is rejected until the font is registered', function () {
    // Proves the family really is unavailable without the host font.
    expect(fn () => new Template(pack_template_requiring(TEST_FAMILY)))
        ->toThrow(Exception::class, TEST_FAMILY);
});

test('a template requiring a registered family compiles', function () {
    Configuration::registerFontPaths([a_font_file()]);

    $template = new Template(pack_template_requiring(TEST_FAMILY), mode: CompilationMode::Development);

    try {
        $svg = $template->export(exportFormat: ExportFormat::svg(), mode: CompilationMode::Development);
        expect($svg)->toContain('<svg');
    } finally {
        $template->cleanup();
    }
});
